packages/protobuf/src/create.ts

```diff
@@ -260,6 +259,26 @@ function createZeroMessage(desc: DescMessage): Message {
   return msg as Message;
 }
 
+/**
+ * Do we need the prototype chain to track field presence?
+ */
+function needsPrototypeChain(desc: DescMessage): boolean {
+  switch (desc.file.edition) {
+    case EDITION_PROTO3:
+      // proto3 always uses implicit presence, we never need the prototype chain.
+      return false;
+    case EDITION_PROTO2:
+      // proto2 never uses implicit presence, we always need the prototype chain.
+      return true;
+    default:
+      // If a message uses scalar or enum fields with explicit presence, we need
+      // the prototype chain to track presence. This rule does not apply to fields
+      // in a oneof group - they use a different mechanism to track presence.
+      return desc.fields.some(
+        (f) => f.presence != IMPLICIT && f.fieldKind != "message" && !f.oneof,
```

Comment:
What does this do for repeated fields? I'm guessing repeated fields return "IMPLICIT" for presence, right? (Pretty sure that's what happens in other runtimes, like C++, Java, Go.) So if a field had only repeated fields, message fields, and oneof fields, it would return "no need for prototype chain", even when the default field presence for the file/message is explicit, right? If so, might we want to apply this test for "proto2" also, just so that we can further minimize the cases where a prototype is used? On the other hand, will it be surprising to users that use editions that sometimes messages use a prototype and sometimes not?

Reply:
Correct, list (repeated) and map fields report presence IMPLICIT.